src/java.base/share/classes/java/lang/VirtualThread.java

```diff
@@ -0,0 +1,1093 @@
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.ForkJoinPool.ForkJoinWorkerThreadFactory;
+import java.util.concurrent.ForkJoinTask;
+import java.util.concurrent.ForkJoinWorkerThread;
+import java.util.concurrent.Future;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import jdk.internal.event.ThreadSleepEvent;
+import jdk.internal.event.VirtualThreadEndEvent;
+import jdk.internal.event.VirtualThreadPinnedEvent;
+import jdk.internal.event.VirtualThreadStartEvent;
+import jdk.internal.event.VirtualThreadSubmitFailedEvent;
+import jdk.internal.misc.CarrierThread;
+import jdk.internal.misc.InnocuousThread;
+import jdk.internal.misc.Unsafe;
+import jdk.internal.vm.Continuation;
+import jdk.internal.vm.ContinuationScope;
+import jdk.internal.vm.StackableScope;
+import jdk.internal.vm.ThreadContainer;
+import jdk.internal.vm.ThreadContainers;
+import jdk.internal.vm.annotation.ChangesCurrentThread;
+import jdk.internal.vm.annotation.JvmtiMountTransition;
+import sun.nio.ch.Interruptible;
+import sun.security.action.GetPropertyAction;
+import static java.util.concurrent.TimeUnit.*;
+
+/**
+ * A thread that is scheduled by the Java virtual machine rather than the operating
+ * system.
+ */
+class VirtualThread extends Thread {
+    private static final Unsafe U = Unsafe.getUnsafe();
+    private static final ContinuationScope VTHREAD_SCOPE = new ContinuationScope("VirtualThreads");
+    private static final ForkJoinPool DEFAULT_SCHEDULER = createDefaultScheduler();
+    private static final ScheduledExecutorService UNPARKER = createDelayedTaskScheduler();
+    private static final int TRACE_PINNING_MODE = tracePinningMode();
+
+    private static final long STATE = U.objectFieldOffset(VirtualThread.class, "state");
+    private static final long PARK_PERMIT = U.objectFieldOffset(VirtualThread.class, "parkPermit");
+    private static final long CARRIER_THREAD = U.objectFieldOffset(VirtualThread.class, "carrierThread");
+    private static final long TERMINATION = U.objectFieldOffset(VirtualThread.class, "termination");
+
+    // scheduler and continuation
+    private final Executor scheduler;
+    private final Continuation cont;
+    private final Runnable runContinuation;
+
+    // virtual thread state, accessed by VM
+    private volatile int state;
+
+    /*
+     * Virtual thread state and transitions:
+     *
+     *      NEW -> STARTED         // Thread.start
+     *  STARTED -> TERMINATED      // failed to start
+     *  STARTED -> RUNNING         // first run
+     *
+     *  RUNNING -> PARKING         // Thread attempts to park
+     *  PARKING -> PARKED          // yield successful, thread is parked
+     *  PARKING -> PINNED          // yield failed, thread is pinned
```

Comment:
```suggestion
     *  PARKING -> PARKED          // parking successful, thread is parked
     *  PARKING -> PINNED          // parking failed, thread is pinned
```

Reply:
The comment was about cont.yield, will change it to make it clearer.